File: mobile/src/api/finance.ts
import { apiClient } from './client';

export interface Category {
  id: string;
  name: string;
  icon: string | null;
  color: string | null;
}

export interface Transaction {
  id: string;
  amount: number;
  currency: string;
  category_id: string | null;
  merchant: string | null;
  description: string | null;
  date: string;
  is_income: boolean;
  raw_text: string | null;
  category?: Category;
}

export interface FinanceStats {
  total_spent: number;
  total_income: number;
  balance: number;
  by_category: {
    name: string;
    amount: number;
    color: string;
    percentage: number;
  }[];
  daily_spending: {
    date: string;
    amount: number;
  }[];
  insights: string[];
}

export interface FinanceReport {
  has_data: boolean;
  period: string;
  summary: {
    income: number;
    spent: number;
    balance: number;
  };
  top_category: {
    name: string;
    amount: number;
    percentage: number;
  };
  critical_day: {
    date: string;
    amount: number;
  };
  insights: Array<{
    type: 'warning' | 'critical' | 'success';
    title: string;
    desc: string;
  }>;
  by_category: Array<{
    name: string;
    amount: number;
    color: string;
  }>;
  message?: string;
}

export const financeApi = {
  async getCategories(): Promise<Category[]> {
    const response = await apiClient.get<Category[]>('/finance/categories');
    return response.data;
  },

  async getTransactions(year?: number, month?: number): Promise<Transaction[]> {
    const params = year && month ? { year, month } : {};
    const response = await apiClient.get<Transaction[]>('/finance/transactions', { params });
    return response.data;
  },

  async getStats(year?: number, month?: number): Promise<FinanceStats> {
    const params = year && month ? { year, month } : {};
    const response = await apiClient.get<FinanceStats>('/finance/stats', { params });
    return response.data;
  },

  async getReport(year: number, month: number): Promise<FinanceReport> {
    const response = await apiClient.get<FinanceReport>(`/finance/report/${year}/${month}`);
    return response.data;
  },

  async addTransaction(tx: Partial<Transaction>) {
    const response = await apiClient.post<Transaction>('/finance/transactions', tx);
    return response.data;
  },
  
  async updateTransaction(id: string, tx: Partial<Transaction>) {
    const response = await apiClient.patch<Transaction>(`/finance/transactions/${id}`, tx);
    return response.data;
  },

  async deleteTransaction(id: string) {
    await apiClient.delete(`/finance/transactions/${id}`);
  }
};

File: backend/app/schemas/finance.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from uuid import UUID

class CategorySchema(BaseModel):
    id: UUID
    name: str
    icon: Optional[str]
    color: Optional[str]

    class Config:
        from_attributes = True

class TransactionCreate(BaseModel):
    amount: float
    currency: str = "EUR"
    category_id: Optional[UUID] = None
    merchant: Optional[str] = None
    description: Optional[str] = None
    date: datetime = datetime.now()
    is_income: bool = False
    raw_text: Optional[str] = None

class TransactionUpdate(BaseModel):
    amount: Optional[float] = None
    category_id: Optional[UUID] = None
    merchant: Optional[str] = None
    description: Optional[str] = None
    date: Optional[datetime] = None
    is_income: Optional[bool] = None

class TransactionSchema(BaseModel):
    id: UUID
    amount: float
    currency: str
    category_id: Optional[UUID]
    merchant: Optional[str]
    description: Optional[str]
    date: datetime
    is_income: bool
    raw_text: Optional[str]
    created_at: datetime
    category: Optional[CategorySchema] = None

    class Config:
        from_attributes = True

class ParseRequest(BaseModel):
    text: str

class FinanceStats(BaseModel):
    total_spent: float
    total_income: float
    balance: float
    by_category: List[dict] # { name, amount, color, percentage }
    daily_spending: List[dict] # { date, amount }
    insights: List[str] # Puntos críticos o consejos

class BudgetSchema(BaseModel):
    id: UUID
    category: str
    limit_percentage: float

class BudgetCreate(BaseModel):
    category: str
    limit_percentage: float
